virtualcluster/pkg/controller/controllers/capi_virtualcluster_controller.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	tenancyv1alpha1 "sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/apis/tenancy/v1alpha1"
+	"sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/controller/constants"
+	"sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/controller/controllers/provisioner"
+	kubeutil "sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/controller/util/kube"
+	strutil "sigs.k8s.io/cluster-api-provider-nested/virtualcluster/pkg/controller/util/strings"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+)
+
+var (
+	_ reconcile.Reconciler = &ReconcileCAPIVirtualCluster{}
+
+	ownerKey = ".metadata.controller"
+	apiGVStr = tenancyv1alpha1.SchemeGroupVersion.String()
+)
+
+// ReconcileCAPIVirtualCluster reconciles a VirtualCluster object
+type ReconcileCAPIVirtualCluster struct {
+	client.Client
+	Log             logr.Logger
+	ProvisionerName string
+	Provisioner     provisioner.Provisioner
+}
+
+// SetupWithManager will configure the VirtualCluster reconciler
+func (r *ReconcileCAPIVirtualCluster) SetupWithManager(mgr ctrl.Manager, opts controller.Options) error {
+	if err := mgr.GetFieldIndexer().IndexField(context.Background(), &clusterv1.Cluster{}, ownerKey, func(rawObj client.Object) []string {
```

Comment:
You are going with the direction that CAPI component is provisioned outside of the vc-manager. Dose this code mean that the clusterv1.cluster object needs to have the owner set to a VC object? If so, where is the workflow to set it?

Reply:
Good catch, I was originally going to add this but then backed away from an owners model and forget to remove these code.

Reply:
We can always add this later if it makes sense, for now with garbage collection we use the finalizer to then trigger deletions of the `Cluster` object as you see later in this package.